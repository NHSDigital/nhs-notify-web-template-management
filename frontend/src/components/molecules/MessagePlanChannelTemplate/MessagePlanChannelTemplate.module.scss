.channel-template {
  &-outer {
    padding: 20px;
    margin-left: 60px;
    background-color: var(--nhsuk-card-background-colour);
    border: 1px solid var(--nhsuk-border-colour);
  }

  &-inner {
    padding: 20px;
    border: 2px solid var(--nhsuk-border-colour);
  }

  &-actions {
    list-style: none;
    margin: 0;
    padding: 0;
  }

  &-link {
    &--remove {
      color: var(--nhsuk-error-colour);

      &:visited {
        color: var(--nhsuk-error-colour);
      }
    }
  }
}
