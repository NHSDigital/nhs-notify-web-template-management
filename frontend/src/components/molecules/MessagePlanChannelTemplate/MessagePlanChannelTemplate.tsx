import Link from 'next/link';
import { Channel, RoutingConfig, TemplateDto } from 'nhs-notify-backend-client';
import {
  channelDisplayMappings,
  channelToTemplateType,
  messagePlanChooseTemplateUrl,
} from 'nhs-notify-web-template-management-utils';
import classNames from 'classnames';

import styles from '@molecules/MessagePlanChannelTemplate/MessagePlanChannelTemplate.module.scss';

import copy from '@content/content';
const { messagePlanChannelTemplate: content } = copy.components;

export function MessagePlanChannelTemplate({
  channel,
  template,
  routingConfigId,
  required = true,
}: {
  channel: Channel;
  routingConfigId: RoutingConfig['id'];
  template?: TemplateDto;
  required?: boolean;
}) {
  const channelDisplayText =
    channelDisplayMappings(channel) + (required ? '' : ` ${content.optional}`);

  return (
    <div className={styles['channel-template-outer']}>
      <div className={styles['channel-template-inner']}>
        <h3 className='nhsuk-heading-s'>{channelDisplayText}</h3>

        {template && template.id && (
          <p
            className='nhsuk-u-margin-bottom-2'
            data-testid={`template-name-${channel}`}
          >
            {template.name}
          </p>
        )}

        <ul className={styles['channel-template-actions']}>
          {!template && (
            <li>
              <Link
                data-testid={`choose-template-link-${channel}`}
                className='nhsuk-link nhsuk-link--no-visited-state'
                href={`/message-plans/${messagePlanChooseTemplateUrl(channelToTemplateType(channel))}/${routingConfigId}`}
              >
                {content.templateLinks.choose}
                <span className='nhsuk-u-visually-hidden'>
                  {channelDisplayText}
                </span>{' '}
                {content.templateLinks.template}
              </Link>
            </li>
          )}

          {template && template.id && (
            <>
              <li>
                <Link
                  data-testid={`change-template-link-${channel}`}
                  className='nhsuk-link nhsuk-link--no-visited-state'
                  href={`/message-plans/${messagePlanChooseTemplateUrl(channelToTemplateType(channel))}/${routingConfigId}`}
                >
                  {content.templateLinks.change}
                  <span className='nhsuk-u-visually-hidden'>
                    {channelDisplayText}
                  </span>{' '}
                  {content.templateLinks.template}
                </Link>
              </li>
              <li>
                <button
                  data-testid={`remove-template-link-${channel}`}
                  className={classNames(
                    styles['channel-template-link--remove'],
                    'nhsuk-link'
                  )}
                  type='button'
                >
                  {content.templateLinks.remove}
                  <span className='nhsuk-u-visually-hidden'>
                    {channelDisplayText}
                  </span>{' '}
                  {content.templateLinks.template}
                </button>
              </li>
            </>
          )}
        </ul>
      </div>
    </div>
  );
}
