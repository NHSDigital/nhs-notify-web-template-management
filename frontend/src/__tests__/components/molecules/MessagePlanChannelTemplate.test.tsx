import { render, screen } from '@testing-library/react';
import { MessagePlanChannelTemplate } from '@molecules/MessagePlanChannelTemplate/MessagePlanChannelTemplate';
import type { TemplateDto } from 'nhs-notify-backend-client';

describe('MessagePlanChannelTemplate', () => {
  it('should display the channel heading', () => {
    render(<MessagePlanChannelTemplate channel="EMAIL" />);

    expect(screen.getByRole('heading', { level: 3, name: 'Email' })).toBeInTheDocument();
  });

  describe('when the channel is not required', () => {
    beforeEach(() => {
      render(<MessagePlanChannelTemplate channel="LETTER" required={false} />);
    })

    it('should display the heading with the "(optional)" suffix', () => {
      expect(
        screen.getByRole('heading', { level: 3, name: 'Letter (optional)' })
      ).toBeInTheDocument();
    });
  });

  describe('when no template is selected', () => {
    beforeEach(() => {
      render(<MessagePlanChannelTemplate channel="NHSAPP" />);
    })

    it('should show the "Choose template" link with accessible name and href', () => {
      const link = screen.getByRole('link', { name: /Choose\s*NHS App\s*template/i });

      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', ''); // TODO: CHANGE
    });

    it('should not display the "Change template" or "Remove template" links', () => {
      expect(
        screen.queryByRole('link', { name: /Change\s*NHS App\s*template/i })
      ).not.toBeInTheDocument();
      expect(
        screen.queryByRole('link', { name: /Remove\s*NHS App\s*template/i })
      ).not.toBeInTheDocument();
    });
  });

  describe('when a template has been selected', () => {
    const testTemplate = { id: 'test-id', name: 'Covid 65+ invitation v1' } as TemplateDto;

    beforeEach(() => {
      render(<MessagePlanChannelTemplate channel="SMS" template={testTemplate} />);
    })

    it('should display the selected template name', () => {
      expect(screen.getByText(testTemplate.name)).toBeInTheDocument();
    });

    it('should display the "Change template" link with accessible name and href', () => {
      const link = screen.getByRole('link', { name: /Change\s*Text message \(SMS\)\s*template/i });

      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', ''); // TODO: CHANGE
    });

    it('should display the "Remove template" link with accessible name and href', () => {
      const link = screen.getByRole('link', { name: /Remove\s*Text message \(SMS\)\s*template/i });

      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', ''); // TODO: CHANGE
    });

    it('should not display the "Choose template" link', () => {
      expect(
        screen.queryByRole('link', { name: /Choose\s*Text message \(SMS\)\s*template/i })
      ).not.toBeInTheDocument();
    });
  });

  it.each(['NHSAPP', 'EMAIL', 'SMS', 'LETTER'] as const)(
    'should match snapshot for empty state (%s)',
    (channel) => {
      const { container } = render(<MessagePlanChannelTemplate channel={channel} />);

      expect(container).toMatchSnapshot();
    }
  );

  it.each(['NHSAPP', 'EMAIL', 'SMS', 'LETTER'] as const)(
    'should match snapshot for selected template state (%s)',
    (channel) => {
      const testTemplate = { id: 'test-id', name: 'Covid 65+ reminder v1' } as TemplateDto;

      const { container } = render(
        <MessagePlanChannelTemplate channel={channel} template={testTemplate} />
      );
      expect(container).toMatchSnapshot();
    }
  );
});
