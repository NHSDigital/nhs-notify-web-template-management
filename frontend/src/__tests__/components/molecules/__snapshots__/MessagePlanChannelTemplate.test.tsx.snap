// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagePlanChannelTemplate should match snapshot for empty state (EMAIL) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        Email
      </h3>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="choose-template-link-EMAIL"
            href="/message-plans/choose-email-template"
          >
            Choose
            <span
              class="nhsuk-u-visually-hidden"
            >
              Email
            </span>
             
            template
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for empty state (LETTER) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        Letter
      </h3>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="choose-template-link-LETTER"
            href="/message-plans/choose-standard-english-letter-template"
          >
            Choose
            <span
              class="nhsuk-u-visually-hidden"
            >
              Letter
            </span>
             
            template
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for empty state (NHSAPP) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        NHS App
      </h3>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="choose-template-link-NHSAPP"
            href="/message-plans/choose-nhs-app-template"
          >
            Choose
            <span
              class="nhsuk-u-visually-hidden"
            >
              NHS App
            </span>
             
            template
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for empty state (SMS) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        Text message (SMS)
      </h3>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="choose-template-link-SMS"
            href="/message-plans/choose-text-message-template"
          >
            Choose
            <span
              class="nhsuk-u-visually-hidden"
            >
              Text message (SMS)
            </span>
             
            template
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for selected template state (EMAIL) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        Email
      </h3>
      <p
        class="nhsuk-u-margin-bottom-2"
        data-testid="template-name-EMAIL"
      >
        Covid 65+ reminder v1
      </p>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="change-template-link-EMAIL"
            href="/message-plans/choose-email-template"
          >
            Change
            <span
              class="nhsuk-u-visually-hidden"
            >
              Email
            </span>
             
            template
          </a>
        </li>
        <li>
          <button
            class="channel-template-link--remove nhsuk-link"
            data-testid="remove-template-link-EMAIL"
            type="button"
          >
            Remove
            <span
              class="nhsuk-u-visually-hidden"
            >
              Email
            </span>
             
            template
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for selected template state (LETTER) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        Letter
      </h3>
      <p
        class="nhsuk-u-margin-bottom-2"
        data-testid="template-name-LETTER"
      >
        Covid 65+ reminder v1
      </p>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="change-template-link-LETTER"
            href="/message-plans/choose-standard-english-letter-template"
          >
            Change
            <span
              class="nhsuk-u-visually-hidden"
            >
              Letter
            </span>
             
            template
          </a>
        </li>
        <li>
          <button
            class="channel-template-link--remove nhsuk-link"
            data-testid="remove-template-link-LETTER"
            type="button"
          >
            Remove
            <span
              class="nhsuk-u-visually-hidden"
            >
              Letter
            </span>
             
            template
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for selected template state (NHSAPP) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        NHS App
      </h3>
      <p
        class="nhsuk-u-margin-bottom-2"
        data-testid="template-name-NHSAPP"
      >
        Covid 65+ reminder v1
      </p>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="change-template-link-NHSAPP"
            href="/message-plans/choose-nhs-app-template"
          >
            Change
            <span
              class="nhsuk-u-visually-hidden"
            >
              NHS App
            </span>
             
            template
          </a>
        </li>
        <li>
          <button
            class="channel-template-link--remove nhsuk-link"
            data-testid="remove-template-link-NHSAPP"
            type="button"
          >
            Remove
            <span
              class="nhsuk-u-visually-hidden"
            >
              NHS App
            </span>
             
            template
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`MessagePlanChannelTemplate should match snapshot for selected template state (SMS) 1`] = `
<div>
  <div
    class="channel-template-outer"
  >
    <div
      class="channel-template-inner"
    >
      <h3
        class="nhsuk-heading-s"
      >
        Text message (SMS)
      </h3>
      <p
        class="nhsuk-u-margin-bottom-2"
        data-testid="template-name-SMS"
      >
        Covid 65+ reminder v1
      </p>
      <ul
        class="channel-template-actions"
      >
        <li>
          <a
            class="nhsuk-link nhsuk-link--no-visited-state"
            data-testid="change-template-link-SMS"
            href="/message-plans/choose-text-message-template"
          >
            Change
            <span
              class="nhsuk-u-visually-hidden"
            >
              Text message (SMS)
            </span>
             
            template
          </a>
        </li>
        <li>
          <button
            class="channel-template-link--remove nhsuk-link"
            data-testid="remove-template-link-SMS"
            type="button"
          >
            Remove
            <span
              class="nhsuk-u-visually-hidden"
            >
              Text message (SMS)
            </span>
             
            template
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
